// @vechain/connex/esm/driver.js
/* eslint-disable @typescript-eslint/no-unused-vars */
import { DriverNoVendor, SimpleNet } from '@vechain/connex-driver';
import { loadLibrary } from './script-loader';
import randomBytes from 'randombytes';
import { blake2b256 } from 'thor-devkit';
const BUDDY_SRC = 'https://unpkg.com/@vechain/connex-wallet-buddy@0.1';
const BUDDY_LIB_NAME = 'ConnexWalletBuddy';
/** the driver implements vendor methods only */
export class DriverVendorOnly {
    constructor(genesisId) {
        this.buddy = loadLibrary(BUDDY_SRC, BUDDY_LIB_NAME).then(lib => lib.create(genesisId, () => randomBytes(16).toString('hex'), val => blake2b256(val).toString('hex')));
    }
    get genesis() { throw new Error('not implemented'); }
    get head() { throw new Error('not implemented'); }
    pollHead() { throw new Error('not implemented'); }
    getBlock(revision) { throw new Error('not implemented'); }
    getTransaction(id, allowPending) { throw new Error('not implemented'); }
    getReceipt(id) { throw new Error('not implemented'); }
    getAccount(addr, revision) { throw new Error('not implemented'); }
    getCode(addr, revision) { throw new Error('not implemented'); }
    getStorage(addr, key, revision) { throw new Error('not implemented'); }
    explain(arg, revision, cacheHints) { throw new Error('not implemented'); }
    filterEventLogs(arg) { throw new Error('not implemented'); }
    filterTransferLogs(arg) { throw new Error('not implemented'); }
    signTx(msg, options) {
        return this.buddy.then(b => b.signTx(msg, options));
    }
    signCert(msg, options) {
        return this.buddy.then(b => b.signCert(msg, options));
    }
}
/** fully implemented Connex.Driver */
class FullDriver extends DriverNoVendor {
    constructor(node, genesis) {
        super(new SimpleNet(node), genesis);
        this.vd = new DriverVendorOnly(genesis.id);
    }
    signTx(msg, options) {
        return this.vd.signTx(msg, options);
    }
    signCert(msg, options) {
        return this.vd.signCert(msg, options);
    }
}
const cache = {};
/**
 * create full driver
 * @param node the url of thor node
 * @param genesis the enforced genesis block
 */
export function createFull(node, genesis) {
    const key = blake2b256(JSON.stringify({
        node,
        genesis
    })).toString('hex');
    let driver = cache[key];
    if (!driver) {
        cache[key] = driver = new FullDriver(node, genesis);
    }
    return driver;
}
