// @vechain/connex/esm/index.js
import { Framework } from '@vechain/connex-framework';
import { genesisBlocks } from './config';
import { compat1 } from './compat';
import { createFull, DriverVendorOnly } from './driver';
import { newVendor } from '@vechain/connex-framework';
/** convert options.network to Connex.Thor.Block */
function normalizeNetwork(n) {
    n = n || 'main';
    if (typeof n === 'string') {
        const gb = genesisBlocks[n];
        if (!gb) {
            throw new Error('invalid network');
        }
        return gb;
    }
    else {
        return n;
    }
}
/** convert network name to genesis id */
function normalizeGenesisId(id) {
    id = id || 'main';
    if (/^0x[0-9a-f]{64}$/.test(id)) {
        return id;
    }
    const gb = genesisBlocks[id];
    if (gb) {
        return gb.id;
    }
    throw new Error('invalid genesis id');
}
/** Vendor class which can work standalone to provides signing-services only */
class VendorClass {
    constructor(genesisId) {
        genesisId = normalizeGenesisId(genesisId);
        try {
            // to detect injected connex
            const injected = (window || {}).connex;
            if (injected && injected.thor.genesis.id === genesisId) {
                // injected genesis id matched
                if (/^1\./.test(injected.version)) {
                    // wrap v1 vendor to v2
                    return compat1(injected).vendor;
                }
                return injected.vendor;
            }
        }
        catch ( /**/_a) { /**/ }
        const driver = new DriverVendorOnly(genesisId);
        const vendor = newVendor(driver);
        return {
            get sign() {
                return vendor.sign.bind(vendor);
            }
        };
    }
}
/** Connex class */
class ConnexClass {
    constructor(opts) {
        const genesis = normalizeNetwork(opts.network);
        if (!opts.noV1Compat) {
            try {
                // to detect injected connex
                const injected = (window || {}).connex;
                if (injected && injected.thor.genesis.id === genesis.id) {
                    // injected genesis id matched
                    if (/^1\./.test(injected.version)) {
                        // wrap v1 to v2
                        return compat1(injected);
                    }
                }
            }
            catch ( /**/_a) { /**/ }
        }
        const driver = createFull(opts.node, genesis);
        const framework = new Framework(driver);
        return {
            get thor() { return framework.thor; },
            get vendor() { return framework.vendor; }
        };
    }
}
ConnexClass.Vendor = VendorClass;
export default ConnexClass;
export { ConnexClass as Connex };
